File: src/main/java/io/github/takenoha/towerdefense/tactical/EmptyTacticalEffectSnapshot.java
package io.github.takenoha.towerdefense.tactical;

import io.github.takenoha.towerdefense.domain.TowerType;

/** Neutral singleton returned when no active tactical build is known. */
public final class EmptyTacticalEffectSnapshot implements TacticalEffectSnapshot {
    public static final EmptyTacticalEffectSnapshot INSTANCE = new EmptyTacticalEffectSnapshot();

    private EmptyTacticalEffectSnapshot() {
    }

    @Override
    public double damageMultiplier(TowerType type, TacticalTargetContext target) {
        return 1.0d;
    }

    @Override
    public double attackIntervalMultiplier(TowerType type, TacticalTargetContext target) {
        return 1.0d;
    }

    @Override
    public double rangeAdd(TowerType type) {
        return 0.0d;
    }

    @Override
    public double areaRadiusMultiplier(TowerType type) {
        return 1.0d;
    }

    @Override
    public int chainCountAdd(TowerType type) {
        return 0;
    }

    @Override
    public double slowStrengthMultiplier(TowerType type) {
        return 1.0d;
    }

    @Override
    public double burnDurationMultiplier(TowerType type) {
        return 1.0d;
    }

    @Override
    public double supportBuffMultiplier() {
        return 1.0d;
    }

    @Override
    public double repairCostMultiplier() {
        return 1.0d;
    }

    @Override
    public double towerDamageTakenMultiplier() {
        return 1.0d;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/EmptyTacticalEffectSnapshot.kt
package io.github.takenoha.towerdefense.tactical

import io.github.takenoha.towerdefense.domain.TowerType

/** Neutral singleton returned when no active tactical build is known. */
class EmptyTacticalEffectSnapshot private constructor() : TacticalEffectSnapshot {
    override fun damageMultiplier(type: TowerType, target: TacticalTargetContext): Double = 1.0

    override fun attackIntervalMultiplier(type: TowerType, target: TacticalTargetContext): Double = 1.0

    override fun rangeAdd(type: TowerType): Double = 0.0

    override fun areaRadiusMultiplier(type: TowerType): Double = 1.0

    override fun chainCountAdd(type: TowerType): Int = 0

    override fun slowStrengthMultiplier(type: TowerType): Double = 1.0

    override fun burnDurationMultiplier(type: TowerType): Double = 1.0

    override fun supportBuffMultiplier(): Double = 1.0

    override fun repairCostMultiplier(): Double = 1.0

    override fun towerDamageTakenMultiplier(): Double = 1.0

    companion object {
        @JvmField
        val INSTANCE: EmptyTacticalEffectSnapshot = EmptyTacticalEffectSnapshot()
    }
}
